File: solutions/fsharp/protein-translation/1/ProteinTranslation.fs
module ProteinTranslation

let translate = function
| "AUG" -> "Methionine"
| "UUU" | "UUC" -> "Phenylalanine" 
| "UUA" | "UUG" -> "Leucine"
| "UCU" | "UCC" | "UCA" | "UCG" -> "Serine"
| "UAU" | "UAC" -> "Tyrosine"
| "UGU" | "UGC" -> "Cysteine"
| "UGG" -> "Tryptophan"
| _ -> ""

let isStop codon = 
    codon = "UAA" || codon = "UAG" || codon = "UGA"

let proteins (rna:string) = 
    let x = rna.Length / 3
    rna 
    |> Seq.splitInto x 
    |> Seq.map System.String
    |> Seq.takeWhile (isStop >> not)
    |> Seq.map(translate)
    |> Seq.toList

File: solutions/csharp/protein-translation/1/ProteinTranslation.cs
using System;
using System.Linq;

public static class ProteinTranslation
{
    public static string[] Proteins(string strand)
    {
        return strand.Chunk(3)
            .Select(chars => new string(chars))
            .TakeWhile(protein => !IsStop(protein))
            .Select(Translate)
            .ToArray();
    }

    static bool IsStop(string s)
        => s is "UAA" or "UAG" or "UGA";

    static string Translate(string protein)
    {
        return protein switch
        {
            "AUG" => "Methionine",
            "UUU" or "UUC" => "Phenylalanine",
            "UUA" or "UUG" => "Leucine",
            "UCU" or "UCC" or "UCA" or "UCG" => "Serine",
            "UAU" or "UAC" => "Tyrosine",
            "UGU" or "UGC" => "Cysteine",
            "UGG" => "Tryptophan",
            _ => throw new InvalidOperationException("")
        };
    }

}